keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+
+For instance, the pod:
+
+```
+metadata:
+  name: nginx
+spec:
+  initContainers:
+  - name: copy-files
+    ...
+    resources:
+      requests:
+        cpu: 100m
+  - name: generate-config
+    ...
+    resources:
+      requests:
+        cpu: 300m
+  containers:
+  - name: proxy
+    ...
+    resources:
+      requests:
+        cpu: 50m
+  - name: sidecar-logging
+    ...
+    resources:
+      requests:
+        cpu: 100m
+```
+
+will be handled by the system in the following ways:
+
+* Is considered to consume `300m` cores of CPU by the scheduler (the node must be able to run the largest init container by itself, which is larger than the sum of the proxy and sidecar containers at runtime)
+* The kubelet will not admit the pod if `(the node's allocatable CPU) - (the sum of all running pods)` is less than `300m`
+* The kubelet will create a cgroup for the pod that expects to get roughly `300m` cores, but the container cgroup created for `copy-files` will request `100m` while that init container is running
+* The quota subsystem will block this pod from being created if there is less than `300m` available.
+
+Once a pod is created, it passes through four high level lifecycle phases as seen by the total Kubernetes system. The first phase is `Pending` - the time before the pod is scheduled to a node. The second phase is `Initializing` - the time between when the pod is scheduled on a node and when all init containers have completed successfully and the `Initalized` condition on the pod is set to `true`. A pod without init containers may only remain in this phase very briefly. The third phase is `Running`, when all init containers have completed, and continues until the pod is deleted or reaches a terminal state of success or failure.  The final state is `Completed` which means the pod has no running containers, all resources have been released or cleaned up, and the pod will never again consume those resources.
+
+Note that Kubernetes is a distributed system and the kubelet owns the pod lifecycle, so the value reported by the system will always slightly trail the actual state of the system.  The lifecycle metric may report a pod is `Pending` but some or all resources may already be allocated on the node, or a `Running` pod may have already reached completion but the Kubelet has not yet reported that to the apiserver.
+
+Future features may alter this model, such as the pod resizing proposal, and those rules will be automatically reflected in the resources handled by the system.
+
+
+### User Stories (optional)
+
+#### Story 1
+
+As an administrator of a Kubernetes cluster, I can easily see the requested resources for the pods on the cluster and compare those to the actual usage of those pods with our instrumentation pipeline.
+
+#### Story 2
+
+As an extender of Kubernetes, my custom requested resources should be easily queryable by an administrator for scheduling without requiring Kubernetes code changes.
```

Comment:
"without having to run third-party components" i.e. KSM is I think also a big value add from a user story perspective.